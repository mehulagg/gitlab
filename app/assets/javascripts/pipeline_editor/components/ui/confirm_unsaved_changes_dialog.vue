<script>
export default {
  props: {
    hasUnsavedChanges: {
      type: Boolean,
      required: true,
    },
  },
  created() {
    window.addEventListener('beforeunload', this.confirmChanges);
  },
  destroyed() {
    window.removeEventListener('beforeunload', this.confirmChanges);
  },
  methods: {
    confirmChanges(e) {
      if (this.hasUnsavedChanges) {
        e.preventDefault();
        e.returnValue = '';
      }
    },
  },
  render: () => null,
};
</script>
