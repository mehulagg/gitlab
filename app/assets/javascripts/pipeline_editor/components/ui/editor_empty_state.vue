<script>
import { GlSprintf } from '@gitlab/ui';
import { s__ } from '~/locale';

export default {
  components: {
    GlSprintf,
  },
  i18n: {
    title: s__('PipelineEditor|Optimize your workflow with CI/CD Pipelines'),
    body: s__(
      'PipelineEditor|Create a new %{codeStart}.gitlab-ci.yml%{codeEnd} file at the root of the repository to get started.',
    ),
  },
  inject: {
    emptyStateIllustrationPath: {
      default: '',
    },
  },
};
</script>
<template>
  <div class="gl-display-flex gl-flex-direction-column gl-align-items-center">
    <img :src="emptyStateIllustrationPath" data-testid="empty-state-illustration" />
    <h1 class="gl-font-size-h1">{{ $options.i18n.title }}</h1>
    <p>
      <gl-sprintf :message="$options.i18n.body" class="gl-display-inline">
        <template #code="{ content }">
          <code>{{ content }}</code>
        </template>
      </gl-sprintf>
    </p>
  </div>
</template>
