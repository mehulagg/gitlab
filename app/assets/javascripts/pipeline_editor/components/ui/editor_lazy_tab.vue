<script>
import { GlTab } from '@gitlab/ui';

/**
 * Wrapper of <gl-tab> to lazily render this tab's content
 * when shown **without dismounting it after**.
 *
 * Once the tab is selected it is permanently set as "not-lazy"
 * so it's contents are not dismounted.
 *
 * Usage:
 *
 * Same as <gl-tab>, for example:
 *
 * <gl-tabs>
 *   <editor-lazy-tab title="Tab 1"> ... </editor-lazy-tab>
 *   <editor-lazy-tab title="Tab 2"> ... </editor-lazy-tab>
 * </gl-tabs>
 */
export default {
  components: {
    GlTab,
  },
  props: {
    lazy: {
      type: Boolean,
      required: false,
      default: true,
    },
  },
  data() {
    return {
      isLazy: this.lazy,
    };
  },
  methods: {
    onClick() {
      this.isLazy = false;
    },
  },
};
</script>
<template>
  <gl-tab :lazy="isLazy" v-bind="$attrs" v-on="$listeners" @click="onClick">
    <slot v-for="slot in Object.keys($slots)" :slot="slot" :name="slot"></slot>
  </gl-tab>
</template>
