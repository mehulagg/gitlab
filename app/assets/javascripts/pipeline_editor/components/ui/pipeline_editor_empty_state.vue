<script>
import { GlButton, GlSprintf } from '@gitlab/ui';
import { __ } from '~/locale';
import glFeatureFlagsMixin from '~/vue_shared/mixins/gl_feature_flags_mixin';

export default {
  components: {
    GlButton,
    GlSprintf,
  },
  i18n: {
    title: __('Optimize your workflow with CI/CD Pipelines'),
    body: __(
      'Create a new %{codeStart}.gitlab-ci.yml%{codeEnd} file at the root of the repository to get started.',
    ),
    btnText: __('Create new CI/CD pipeline'),
  },
  mixins: [glFeatureFlagsMixin()],
  inject: {
    emptyStateIllustrationPath: {
      default: '',
    },
  },
  computed: {
    showCTAButton() {
      return this.glFeatures.pipelineEditorEmptyStateAction;
    },
  },
  methods: {
    createEmptyConfigFile() {
      this.$emit('createEmptyConfigFile');
    },
  },
};
</script>
<template>
  <div class="gl-display-flex gl-flex-direction-column gl-align-items-center gl-mt-11">
    <img :src="emptyStateIllustrationPath" />
    <h1 class="gl-font-size-h1">{{ $options.i18n.title }}</h1>
    <p class="gl-mt-3">
      <gl-sprintf :message="$options.i18n.body">
        <template #code="{ content }">
          <code>{{ content }}</code>
        </template>
      </gl-sprintf>
    </p>
    <gl-button
      v-if="showCTAButton"
      variant="confirm"
      class="gl-mt-3"
      @click="createEmptyConfigFile"
    >
      {{ $options.i18n.btnText }}
    </gl-button>
  </div>
</template>
