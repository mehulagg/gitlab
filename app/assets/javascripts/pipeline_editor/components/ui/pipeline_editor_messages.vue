<script>
import { GlAlert } from '@gitlab/ui';
import { getParameterValues, removeParams } from '~/lib/utils/url_utility';
import { __, s__ } from '~/locale';
import {
  COMMIT_FAILURE,
  COMMIT_SUCCESS,
  DEFAULT_FAILURE,
  LOAD_FAILURE_UNKNOWN,
} from '../../constants';
import CodeSnippetAlert from '../code_snippet_alert/code_snippet_alert.vue';
import {
  CODE_SNIPPET_SOURCE_URL_PARAM,
  CODE_SNIPPET_SOURCES,
} from '../code_snippet_alert/constants';

export default {
  components: {
    GlAlert,
    CodeSnippetAlert,
  },
  errorTexts: {
    [COMMIT_FAILURE]: s__('Pipelines|The GitLab CI configuration could not be updated.'),
    [DEFAULT_FAILURE]: __('Something went wrong on our end.'),
    [LOAD_FAILURE_UNKNOWN]: s__('Pipelines|The CI configuration was not loaded, please try again.'),
  },
  successTexts: {
    [COMMIT_SUCCESS]: __('Your changes have been successfully committed.'),
  },
  props: {
    failureType: {
      type: String,
      required: false,
      default: null,
    },
    failureReasons: {
      type: Array,
      required: false,
      default: () => [],
    },
    successType: {
      type: String,
      required: false,
      default: null,
    },
  },
  data() {
    return {
      codeSnippetCopiedFrom: '',
      showAlert: false,
    };
  },
  computed: {
    alerts() {
      const messages = [];

      if (this.failureType) {
        messages.push(this.failure);
      }

      if (this.success) {
        messages.push(this.success);
      }
      return messages;
    },
    failure() {
      switch (this.failureType) {
        case LOAD_FAILURE_UNKNOWN:
          return {
            text: this.$options.errorTexts[LOAD_FAILURE_UNKNOWN],
            variant: 'danger',
          };
        case COMMIT_FAILURE:
          return {
            text: this.$options.errorTexts[COMMIT_FAILURE],
            variant: 'danger',
          };
        default:
          return {
            text: this.$options.errorTexts[DEFAULT_FAILURE],
            variant: 'danger',
          };
      }
    },
  },
  success() {
    switch (this.successType) {
      case COMMIT_SUCCESS:
        return {
          text: this.$options.successTexts[COMMIT_SUCCESS],
          variant: 'info',
        };
      default:
        return null;
    }
  },

  created() {
    this.parseCodeSnippetSourceParam();
  },
  methods: {
    dismissAlert() {
      this.showAlert = false;
    },
    dismissCodeSnippetAlert() {
      this.codeSnippetCopiedFrom = '';
    },
    parseCodeSnippetSourceParam() {
      const [codeSnippetCopiedFrom] = getParameterValues(CODE_SNIPPET_SOURCE_URL_PARAM);
      if (codeSnippetCopiedFrom && CODE_SNIPPET_SOURCES.includes(codeSnippetCopiedFrom)) {
        this.codeSnippetCopiedFrom = codeSnippetCopiedFrom;
        window.history.replaceState(
          {},
          document.title,
          removeParams([CODE_SNIPPET_SOURCE_URL_PARAM]),
        );
      }
    },
  },
};
</script>
<template>
  <div>
    <code-snippet-alert
      v-if="codeSnippetCopiedFrom"
      :source="codeSnippetCopiedFrom"
      class="gl-mb-5"
      @dismiss="dismissCodeSnippetAlert"
    />
    <gl-alert
      v-for="alert in alerts"
      :key="alert.text"
      :variant="alert.variant"
      class="gl-mb-5"
      @dismiss="dismissAlert"
    >
      {{ alert.text }}
      <ul v-if="failureReasons.length" class="gl-mb-0">
        <li v-for="reason in failureReasons" :key="reason">{{ reason }}</li>
      </ul>
    </gl-alert>
  </div>
</template>
