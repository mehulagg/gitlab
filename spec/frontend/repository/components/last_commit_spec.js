import { shallowMount } from '@vue/test-utils';
import { GlLoadingIcon } from '@gitlab/ui';
import LastCommit from '~/repository/components/last_commit.vue';
import UserAvatarLink from '~/vue_shared/components/user_avatar/user_avatar_link.vue';

let wrapper;

function createCommitData(data = {}) {
  return {
    sha: '123456789',
    title: 'Commit title',
    message: 'Commit message',
    webUrl: 'https://test.com/commit/123',
    authoredDate: '2019-01-01',
    author: {
      name: 'Test',
      avatarUrl: 'https://test.com',
      webUrl: 'https://test.com/test',
    },
    pipeline: {
      detailedStatus: {
        detailsPath: 'https://test.com/pipeline',
        icon: 'failed',
        tooltip: 'failed',
        text: 'failed',
        group: {},
      },
    },
    ...data,
  };
}

function factory(commit = createCommitData(), loading = false) {
  wrapper = shallowMount(LastCommit, {
    mocks: {
      $apollo: {
        queries: {
          commit: {
            loading: true,
          },
        },
      },
    },
    sync: false,
    attachToDocument: true,
  });
  wrapper.setData({ commit });
  wrapper.vm.$apollo.queries.commit.loading = loading;
}

describe('Repository last commit component', () => {
  afterEach(() => {
    wrapper.destroy();
  });

  it.each`
    loading  | label
    ${true}  | ${'shows'}
    ${false} | ${'hides'}
  `('$label when loading icon $loading is true', ({ loading }) => {
    factory(createCommitData(), loading);

    return wrapper.vm.$nextTick(() => {
      expect(wrapper.find(GlLoadingIcon).exists()).toBe(loading);
    });
  });

  it('renders commit widget', () => {
    factory();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders short commit ID', async () => {
    factory();

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.label-monospace').text()).toEqual('12345678');
  });

  it('hides pipeline components when pipeline does not exist', () => {
    factory(createCommitData({ pipeline: null }));

    expect(wrapper.find('.js-commit-pipeline').exists()).toBe(false);
  });

  it('renders pipeline components', async () => {
    factory();

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.js-commit-pipeline').exists()).toBe(true);
  });

  it('hides author component when author does not exist', () => {
    factory(createCommitData({ author: null }));

    expect(wrapper.find('.js-user-link').exists()).toBe(false);
    expect(wrapper.find(UserAvatarLink).exists()).toBe(false);
  });

  it('does not render description expander when description is null', () => {
    factory(createCommitData({ description: null }));

    expect(wrapper.find('.text-expander').exists()).toBe(false);
    expect(wrapper.find('.commit-row-description').exists()).toBe(false);
  });

  it('expands commit description when clicking expander', async () => {
    factory(createCommitData({ description: 'Test description' }));

    await wrapper.vm.$nextTick();

    wrapper.find('.text-expander').trigger('click');

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.commit-row-description').isVisible()).toBe(true);
    expect(wrapper.find('.text-expander').classes('open')).toBe(true);
  });

  it('renders the signature HTML as returned by the backend', () => {
    factory(createCommitData({ signatureHtml: '<button>Verified</button>' }));

    expect(wrapper.element).toMatchSnapshot();
  });
});
