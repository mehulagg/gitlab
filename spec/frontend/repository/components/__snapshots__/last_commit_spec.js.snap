// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository last commit component renders commit widget 1`] = `
<div
  class="info-well d-none d-sm-flex project-last-commit commit p-3"
>
  <glloadingicon-stub
    class="m-auto"
    color="orange"
    label="Loading"
    size="md"
  />
</div>
`;

exports[`Repository last commit component renders the signature HTML as returned by the backend 1`] = `
<div
  class="info-well d-none d-sm-flex project-last-commit commit p-3"
>
  <glloadingicon-stub
    class="m-auto"
    color="orange"
    label="Loading"
    size="md"
  />
</div>
`;
